### ISSUE TITLE
Input.is_action_just_pressed is true on two frames with an InputEventMouseButton

### ISSUE THREAD

**Issue by Rubonnek** [godotengine/godot#43492]
**Godot version:**

3.2 branch at commit adf2dabbde23fdd89803871e26dd1d33ca20bf77

**OS/device including version:**

Arch Linux

**Issue description:**

`Input.is_action_just_pressed` returns true on two frames for the same action with an `InputEventMouseButton` which is unexpected as it should only be true on one frame.

**Steps to reproduce:**
1. Run the attached project
2. Start moving your mouse inside the project window (don't stop moving it).
3. Keep moving the mouse and right clicking the project window and holding the right click for more than two frames.
4. On some occasions `Input.is_action_just_pressed` will be true for two frames.

**Minimal reproduction project:**

[rbb.zip](https://github.com/godotengine/godot/files/5534679/rbb.zip)

**Comment by Calinou** [comment]
Does this occur if you call `Input.set_use_accumulated_input(false)` in any script's `_ready()` function?

**Comment by Rubonnek** [comment]
Yes. In fact, running `Input.set_use_accumulated_input(false)` makes it worse. I'm getting up 3 frames where `Input.is_action_just_pressed` is true.

**Comment by rosshadden** [comment]
I encountered this as well and worked around it using the `event.is_action_type` in my `_input` functions to return early if it is not an action event. This wouldn't work in situations where one `_input` handler handles both action events and mouse events, so duck checking the event type is probably a more generally applicable workaround.

**Comment by Rubonnek** [comment]
@rosshadden is right. In this particular case, the root cause of this issue is simply that I was using the `Input` singleton which has access to all the events that are happening in the current frame. I'm not inspecting the `InputEvent` that is actually handed to the `_gui_input` function and by moving the mouse and right clicking at the same time two `InputEvent` objects will be passed to the `_gui_input` function at the same frame, which in turn will make `Input.is_action_just_pressed("right_click_bug")` return true twice because we are still in the same frame where the right click was pressed and when the mouse was moved.

This can be proved by using the event just as @rosshadden suggested.

Printing the `InputEvent` as follows confirms this is the case as well:

```gdscript
func _gui_input(_p_input_event : InputEvent) -> void:
	if Input.is_action_just_pressed("right_click_bug"):
		count += 1
		message = "Number of right clicks detected: " + str(count)
		print(message)
		$Label.set_text("Number of right clicks detected: " + str(count))
		print(_p_input_event)

```

The following output can be gathered under the same frame:

```
Number of right clicks detected: 1
[InputEventMouseButton:1807]
Number of right clicks detected: 2
[InputEventMouseMotion:1810]
```

By the time I opened this ticket I wasn't aware that any of the input related functions (i.e. `_input`, `_unhandled_input`, `__unhandled_key_input`, and `_gui_input`) could fire more than once under the same frame. And in general, using the `Input` singleton within any of these function could be considered bad practice (if used, it should be used carefully on special cases where more than one input needs to be inspected, just like @rosshadden pointed out. I'll be opening a new ticket to document this. I think this should be or could be included in our best practices guide.

I'm closing this ticket since there is no bug here.
